library/Backends/Plonk/IR.md

```diff
@@ -0,0 +1,103 @@
+
+## Binary operations
+
+| Opcode | Opcode Bytes    | Types | Description     | Notes |
+| ------ | --------------- | ----- | --------------- | ----- |
+| add    | `[a], [b], [o]` |       | o = a + b       |       |
+| sub    | `[a], [b], [o]` |       | o = a - b       |       |
+| mul    | `[a], [b], [o]` |       | o = a * b       |       |
+| div    | `[a], [b], [o]` |       | o = a / b       |       |
+| mod    | `[a], [b], [o]` |       | o = a % b       |       |
+| and    | `[a], [b], [o]` |       | `o = a &amp; b` |       |
+| or     | `[a], [b], [o]` |       | `o = a | b`     |       |
+| xor    | `[a], [b], [o]` |       | `o = a ^ b`     |       |
+
+```haskell
+data BinOp f a where
+  BAdd :: BinOp f f
+  BSub :: BinOp f f
+  BMul :: BinOp f f
+  BDiv :: BinOp f f
+  BMod :: BinOp f f
+  BExp :: BinOp f f
+  BAnd :: BinOp f Bool
+  BOr :: BinOp f Bool
+  BXor :: BinOp f Bool
+```
+
+## Comparing operations
+
+These are also binary operations with the constraint that the resulting value is of type `Boolean`.
+
+| Opcode | Opcode Bytes    | Description       | Notes |
+| ------ | --------------- | ----------------- | ----- |
+| gt     | `[a], [b], [o]` | `o = (a &gt; b)`  |       |
+| gte    | `[a], [b], [o]` | `o = (a &gt;= b)` |       |
+| lt     | `[a], [b], [o]` | `o = (a &lt; b)`  |       |
+| lte    | `[a], [b], [o]` | `o = (a &lt;= b)` |       |
+| eq     | `[a], [b], [o]` | `o = (a == b)`    |       |
+
+```haskell
+data CompOp f where
+  CGt :: CompOp f
+  CGte :: CompOp f
+  CLt :: CompOp f
+  CLte :: CompOp f
+  CEq :: CompOp f
+```
+
+## Unary operations
+
+| Opcode   | Opcode Bytes  | Description                    | Notes                 |
+| -------- | ------------- | ------------------------------ | --------------------- |
+| setpub   | `[p], [o]`    | o = p                          | from public calldata  |
+| setpriv  | `[p], [o]`    | o = p                          | from private calldata |
+| dup      | `[a], [o]`    | `o = a`                        |                       |
+| iszero   | `[a], [o]`    | `o = (a == 0)`                 |                       |
+| not      | `[a], [o]`    | `o = !a`                       |                       |
+| shl      | `[a], x, [o]` | `o = a &lt;&lt; x`             | x is const            |
+| shr      | `[a], x, [o]` | `o = a &gt;&gt; x`             |                       |
+| rotl     | `[a], x, [o]` | `o = a &lt;&lt;&lt; x`         |                       |
+| rotr     | `[a], x, [o]` | `o = a &gt;&gt;&gt; x`         |                       |
+| asserteq | `[a], [o]`    | throw if `a != 0`              |                       |
+| assertlt | `[a], x`      | throw if `a &gt;= 2^x`         |                       |
+| mload    | `[x], [o]`    | read `o` from RAM location `x` |                       |
+| mstore   | `[x], [a]`    | write `a` to RAM location `x`  |                       |
+
+```haskell
+data UnOp f a where
```

Comment:
Seems we're missing `setpub` and `setpriv` and the memory operations, is that intentional?